### DISCUSSION TITLE
AI deck cannot be flashed with wifi example (old bootloader, flashing freezes)

### DISCUSSION THREAD

**Discussion by boomer319** [bitcraze/discussions#942] [Q&A] [answered]
Hello,

I am working with multiple crazyflies 2.1 and their respective AI decks (Ver. 1.1 Rev D). I am runnning Ubuntu 22.04.

I was following this tutorial (https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) ... downloaded aideck_gap8_wifi_img_streamer_with_ap.bin ... tried to flash it via the following command:

cfloader flash aideck_gap8_wifi_img_streamer_with_ap.bin deck-bcAI:gap8-fw -w radio://0/80/2M/E7E7E7E7E7

The flashing starts, goes on and freezes at 99%:

Reset to bootloader mode ...
Could not save cache, no writable directory
Could not save cache, no writable directory
Skipping bcAI:esp, not in the target list
Deck bcAI:gap8, reset to bootloader
| 0% Writing to bcAI:gap8 deck memory
/ 1% Writing to bcAI:gap8 deck memory
o 2% Writing to bcAI:gap8 deck memory
\ 3% Writing to bcAI:gap8 deck memory
...

The version of the bootloader on certain AI decks (Version 1.1 Revision D) is old, making the device not usable for (any?) applications out of the box and to flash anything to the AI deck via the CrazyRadio one must first flash the new bootloader onto the AI decks system via a special connector on the AI deck (JTAG).

Partly stated here: https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader

> Note, this requires the GAP8 bootloader to be flashed on the GAP8 (which is the case for all AIdecks equal or newer than version 1.1 Rev D).

This is stated here: https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/flashing/

Unfortunately, in our lab we do not have the JTAG debugging adapters specified in the tutorials about flashing the new bootloader.

However we ordered the debugging kit. Even though it was mentioned in the discussion linked below, I do not see how it should be used to connect to the GAP8 debugger, with the accesories in the box at least.

In the forums I found this discussion (https://github.com/orgs/bitcraze/discussions/823) saying a J-link or an ARM-USB-TINY-H JTAG debug probe would be needed. However, the models I see on the internet don't seem to have the plug to fit the GAP8 debugging connector and prices vary from 8$ to multiple hundreds of dollars, disrupting the decision making process.

The application I would use the AI deck for would be obstacle avoidance. I assume, to make use of the AI deck in any way I would need to first get a fitting JTAG debugger and flash a new bootloader.

Is there an inexpensive option for a debugger (and possibly adapter combo) you have tested yourself, meaning: that is known to work?

To be sure:
Do I even need to flash the new bootloader on the AI decks system to make it functional for any use case? Is the necessary configuration already on the AI deck?

Best regards

**Comment by tobbeanton** [comment]
> The version of the bootloader on certain AI decks (Version 1.1 Revision D) is old, making the device not usable for (any?) applications out of the box and to flash anything to the AI deck via the CrazyRadio one must first flash the new bootloader onto the AI decks system via a special connector on the AI deck (JTAG).

Yes that is true, there is a note in the [product page](https://www.bitcraze.io/products/ai-deck/) and in the [shop](https://store.bitcraze.io/collections/decks/products/ai-deck-1-1) but perhaps it can be clearer? Also it will be hard to do any advanced development without a JTAG, even with a working bootloader, so I would say it is a requirement.

> The application I would use the AI deck for would be obstacle avoidance. I assume, to make use of the AI deck in any way I would need to first get a fitting JTAG debugger and flash a new bootloader.
Yes and also for convenience when developing.

I see now that the some of the [JTAG programming documentation](https://github.com/bitcraze/aideck-gap8-examples/blob/508b1938ae4476bc2679ef2dd732254dcecf1f5c/docs/getting-started/jtag-programming.md) has fallen away when its been refactored. There you find the adapter information. We will restore what is needed from this. Thanks for finding it!

> To be sure:
Do I even need to flash the new bootloader on the AI decks system to make it functional for any use case? Is the necessary configuration already on the AI deck?

Yes. (the wifi camera example flashed on there in production is not up to date anymore and I'm not sure the pc-python side works with it anymore).

**Reply by boomer319** [reply]
Very nice! Thank you for laying it out so clearly and linking to your [documentation](https://github.com/bitcraze/aideck-gap8-examples/blob/508b1938ae4476bc2679ef2dd732254dcecf1f5c/docs/getting-started/jtag-programming.md).

Seeing that ESP32 programmers work too: Would this one work as well ([ESP32 Programmer/Debugger](https://www.digikey.ca/en/products/detail/espressif-systems/ESP-PROG/10259352))? Will I be able to follow the "software" steps shown in your tutorials with this generic esp32 debugger?
e.g., [this](https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/) tutorial:

>$ git clone https://github.com/bitcraze/aideck-gap8-bootloader.git
$ cd aideck-gap8-bootloader
$ docker run --rm -it -v $PWD:/module/ --device /dev/ttyUSB0 --privileged -P bitcraze/aideck /bin/bash -c 'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'

I would assume the
>/dev/ttyUSB0

and the

>'export GAPY_OPENOCD_CABLE=interface/ftdi/olimex-arm-usb-tiny-h.cfg; source /gap_sdk/configs/ai_deck.sh; cd /module/;  make all image flash'

must be changed. Must they actually be changed? If yes to what values...since the olimex-arm-usb-tiny-h.cfg is a config file for the specific debugger?

**Reply by knmcguire** [reply]
Hi! yes you will need to `change interface/ftdi/olimex-arm-usb-tiny-h.cfg`. This is an configuration file of openocd that need to be changed to your right debugger.

I'm actually not sure if the esp debugger is supposed to be in that list. We have only tried with jlinks and olimex, but I remember an error from the esp-debugger as it was missing a necessary reset pin... let me double check this. I think it was only suitable for the Nina module reflash but not for the aideck.

**Reply by tobbeanton** [reply]
Unfortunately the ESP32 programmer didn't work as @knmcguire pointed out (or at least not without HW patching and changing the OpenOCD config. Don't remember the details). I do believe any FTDI based JTAG programmer will work though and perhaps there are other alternatives then the Olimex.

**Reply by knmcguire** [reply, marked as answer]
@boomer319 I noticed that the documentation you referred two before (this one https://github.com/bitcraze/aideck-gap8-examples/blob/508b1938ae4476bc2679ef2dd732254dcecf1f5c/docs/getting-started/jtag-programming.md) is about 2 years old. In the mean time we removed the esp debugger from the list. 

We should add the jlinks to it as well, at least the none edu mini version but the normal one works as I've been using that all the time.

**Reply by boomer319** [reply]
Thanks a lot!

I have found two types of jlinks: The [AliExpress copies for 40$](https://www.aliexpress.com/item/1005004471558568.html) and [Digikey, most likely, originals for ~750$](https://www.digikey.ca/en/products/detail/segger-microcontroller-systems/8.08.00/2175882)

Since I don't know if I can trust the copies on AliExpress (most likely little to no documentation available), I will probably go with the [Olimex (~60$)](https://www.digikey.ca/en/products/detail/olimex-ltd/ARM-USB-TINY-H/3471388) and its [adapter (~8$)](https://www.digikey.ca/en/products/detail/olimex-ltd/ARM-JTAG-20-10/3471401).

Will that work out?
Has anyone had experiences with the shady AliExpress copies?

**Reply by knmcguire** [reply]
Not us personally in the bitcraze office... perhaps one of these days we should try it out but yes it is a bit risky.

**Reply by boomer319** [reply]
I got the Olimex and the adapter. Flashing works beautifully. Thanks!

**Comment by decock918** [comment]
The version of the bootloader on certain AI decks (Version 1.1 Revision D) is old, making the device not usable for (any?) applications out of the box and to flash anything to the AI deck via the CrazyRadio one must first flash the new bootloader onto the AI decks system via a special connector on the AI deck (JTAG).
[pikashow download
](https://pikashow.fyi/get-apk/
)

**Reply by knmcguire** [reply]
Hi! Technically it should only be the case with revision C or down (not D) to reflash the gap8 bootloader but some production errors have caused issues with the newer boards as well. None the less, **always check if over-air-flashing** works before reflashing.

https://www.bitcraze.io/documentation/tutorials/getting-started-with-aideck/#gap8-bootloader
